src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs

```diff
@@ -111,52 +112,95 @@ public virtual bool ShouldFormatSnippet(SnippetInfo snippetInfo)
 
         private void PopulateSnippetCaches()
         {
-            var updatedSnippets = GetSnippetInfoList();
-            var updatedSnippetShortcuts = GetShortcutsHashFromSnippets(updatedSnippets);
+            Debug.Assert(_expansionManager != null);
 
-            lock (cacheGuard)
-            {
-                snippets = updatedSnippets;
-                snippetShortcuts = updatedSnippetShortcuts;
-            }
-        }
+            // Ideally we'd fork execution here based on whether the expansion manager is an
+            // IExpansionManager or not. Unfortunately, we cannot mention that type by name until
+            // the Roslyn build machines are upgraded to Visual Studio 2015 Update 1. We therefore
+            // need to try using IExpansionManager dynamically, from a background thread. If that
+            // fails, then we come back to the UI thread for using IVsExpansionManager instead.
 
-        protected static HashSet<string> GetShortcutsHashFromSnippets(IList<SnippetInfo> updatedSnippets)
-        {
-            return new HashSet<string>(updatedSnippets.Select(s => s.Shortcut), StringComparer.OrdinalIgnoreCase);
+            var token = _waiter.BeginAsyncOperation(GetType().Name + ".Start");
+
+            Task.Factory.StartNew(async () => await PopulateSnippetCacheOnBackgroundWithForegroundFallback().ConfigureAwait(false),
+                            CancellationToken.None,
+                            TaskCreationOptions.None,
+                            TaskScheduler.Default).CompletesAsyncOperation(token);
         }
 
-        private IList<SnippetInfo> GetSnippetInfoList()
+        private async Task PopulateSnippetCacheOnBackgroundWithForegroundFallback()
         {
-            IVsExpansionEnumeration expansionEnumerator;
-            if (TryGetVsSnippets(out expansionEnumerator))
+            AssertIsBackground();
+
+            try
+            {
+                IVsExpansionEnumeration expansionEnumerator = await ((dynamic)_expansionManager).EnumerateExpansionsAsync(
+                    _languageGuidForSnippets,
+                    0, // shortCutOnly
+                    Array.Empty<string>(), // types
+                    0, // countTypes
+                    1, // includeNULLTypes
+                    1 // includeDulicates: Allows snippets with the same title but different shortcuts
+                    ).ConfigureAwait(false);
+
+                // The rest of the process requires being on the UI thread
```

Comment:
@tinaschrepfer Just looked at this with @jasonmalinowski, and I'm going to attempt to explain why I need to be on the UI thread here.

We call `IVsExpansionEnumeration.Next`, which is defined like this:

```C#
int Next([ComAliasName("Microsoft.VisualStudio.OLE.Interop.ULONG")] uint celt, [ComAliasName("Microsoft.VisualStudio.TextManager.Interop.VsExpansion")] IntPtr[] rgelt, [ComAliasName("Microsoft.VisualStudio.OLE.Interop.ULONG")] out uint pceltFetched);
```

And we pass a pointer for `rgelt` that we expect to be populated as the result. This eventually calls into the `CExpansionEnumeratorShim::Next` method, which has the same contract of expecting a non-null `rgelt` that it can drop expansion data into. When we call from the UI thread, this transition from managed to the `CExpansionEnumeratorShim` goes smoothly and everything works.

When we call from a background thread, the com marshaller has to move execution to the UI thread, and as part of this process it uses the interface as defined in the idl to set up the appropriate arguments to pass. The idl definition looks like this:

```
HRESULT     Next (ULONG celt, [out, size_is(celt), length_is(*pceltFetched)] VsExpansion **rgelt, [out]ULONG *pceltFetched);
```

Because rgelt is specified as an `out` parameter, we think the marshaller is discarding the pointer we passed and substituting the null reference. This would then cause the failure I'm currently seeing, where the first access of rgelt in the shim is causing the null ref.